aten/src/ATen/core/jit_type.h

```diff
@@ -97,52 +97,87 @@ struct SingleElementType : public Type {
   TypePtr elem;
 };
 
+struct NoneType;
+using NoneTypePtr = std::shared_ptr<NoneType>;
+// This type represents a Python None
+struct TORCH_API NoneType : public Type {
+  static NoneTypePtr create() {
+    return NoneTypePtr(new NoneType()); // NOLINT(modernize-make-shared)
+  }
+  bool operator==(const Type& rhs) const override {
+    return rhs.kind() == kind();
+  }
+  std::string str() const override {
+    return "None";
+  }
+  bool isSubtypeOfExt(const TypePtr& rhs, std::ostream *why_not) const override;
+
+  static const TypeKind Kind = TypeKind::NoneType;
+  // global singleton
+  static NoneTypePtr get();
+
+ private:
+  NoneType() : Type(TypeKind::NoneType) {}
+};
+
+
+template <TypeKind K>
+// This represents any type that we can dynamically specialize at
+// runtime. Currently only `Optional` and `Union` inherit from
+// `DynamicType`
+struct DynamicType : public Type {
```

Comment:
`Optional` used to inherit from `SingleElementType`, which exposed certain methods that I felt made sense in `Optional` but didn't make sense in `Union`. Also, `Union` has certain methods that didn't make sense in `Optional` as Optional is CURRENTLY used in the codebase. 

I have plans to change this later, though. It will be possible in my follow-up project where I remove all internal instances of `Optional` and make `Optional[T]` a simple alias for `Union[T, None]` [the way Python does](https://github.com/python/typing/blob/master/src/typing.py#L850).

Reply:
Thanks for the feedback, changing now!